### DISCUSSION TITLE
Loaders

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#454] [Proposal]
### Body

https://vimeo.com/795022419/fe90a97420

# Summary

A new package that provides APIs for creating *loaders*, functions that fetch data from an external source. The loaders package provides low-level APIs for building custom loaders and higher-level APIs for common scenarios.

# Background & Motivation

A common issue when working with Astro is that fetching external data can some times result in overfetching. Users have reported problems such as data being different between pages in an SSG build, and exceeding a rate-limit during dev. The loaders package is intended to solve these issues.

# Goals

- Provide high-level solutions for common overfetching provides such as:
  - Exceeding rate limits in APIs such as GitHub's.
  - Unwanted data differences in an SSG build caused by the fact that pages are generated at different times.
- Provide low-level APIs that can be used to build custom loaders. People are opinionated on caching and have different needs; providing a low-level base allows us to provide some higher-level APIs that are slightly opinionated without preventing people from doing what they need.
- Account for use-cases where there are multiple caching layers. For example, a user might want to do multiple of these:
  - Checking an in-memory cache layer such as Redis or Memcache.
  - Checking an in-process LRU cache.
  - Using a fall-through cache, where a cached response is provided and the cache is updated in the background for the next request.
- Allow loaders to be used in any framework or non-framework code; do not limit to Astro files.
- Provide a way to do schema based validation on the API response.

# Non-goals

- Shimming global environments or providing a generic `fetch` that's always available.
- Mimicking the caching behaviors of other frameworks. We will reference other implementations for inspiration but provide a solution that fits the ethos of Astro.
- Having many built-in caching behaviors. We might have a few for common scenarios like rate limited APIs, but will let the community build out many more use-cases.

# Example

This branch has a proof of concept which you can check out here: https://github.com/withastro/astro/tree/loaders/packages/loaders

## Base loader creation

Here's an example of creating a generic loader using the core `createLoader` API:

```js
import { createLoader } from '@astrojs/loaders';

const getItem = createLoader({
  key: id => id,
  async load(id) { return 22; }
});

const value = await getItem('key');
console.log(value); // 22
```

## Using a caching layer

Here's another example that uses the built-in `RateLimitCache` layer:

```js
import { createLoader, RateLimitCache } from '@astrojs/loaders';

const cache = new RateLimitCache({
  allowed: 1000,
  period: 'hour'
});

const getItem = createLoader({
  key: id => id,
  async load(id) {
    let res = await fetch(`http://github.com/api/issues/${id}`);
    return await res.json();
  },
  layers: [cache]
});

for(let _ of [...Array(2000).keys()]) {
  getItem(1);
}
```

In this example the built-in `RateLimitCache` is created, setting itself to 1000 requests per hour. Then we do 2000 requests in parallel.

The cache layer will start returning from the cache on the one thousand and first request.

## Providing a schema

Responses can be validated using a Zod schema, like so:

```js
import { z } from "zod";
import { createLoader } from '@astrojs/loaders';

const getIssue = createLoader({
  key: id => id,
  async load(id) {
    let res = await fetch(`http://github.com/api/issues/${id}`);
    return await res.json();
  },
  schema: z.object({
    id: z.number(),
    title: z.string()
  })
});
```

## High-level `fetch` loader

The loaders package will provide a `fetch` implementation that mimicks the `fetch` API. It provides caching underneath using the `createLoader` API. This is transparent to the user.

The fetch implementation has the following caching behaviors:

- During the build any requests to the same URL will return the same response. This ensures that multiple pages doing the same request will get the same result.
- __?__ That's the only behavior for now. Another idea I had was to speed up slow APIs by using a fall-through cache during dev. This would mean you get an immediate response when refreshing the browser but the request still occurs.

```js
import { fetch } from '@astrojs/loaders';

let response = await fetch('http://example.com/some/api');
let json = await response.json();
```

Additionally there is a `createFetch` export that allows you to pass in your own caching layers:

```js
import { createFetch, BuildCache, RateLimitCache } from '@astrojs/loaders';

let fetch = createFetch(new BuildCache(), new RateLimitCache({
  allowed: 1000,
  period: 'hour'
}));
let response = await fetch('http://example.com/some/api');
let json = await response.json();
```

## Composing loaders

Loaders compose naturally. You can create a loader and use it inside of the `load` function of another. You'll get caching behaviors from both.

## Creating a cache

The cache API is simple (maybe too simple at the moment). It is an object with this interface:

```ts
interface CacheLayer<K> {
  get(id: K): any | Promise<any>;
  cache(id: K, value: unknown): void | Promise<void>;
}
```

Where `K` is the key. A cache layer might only support certain types of keys and can control that with TS `extends`, for example: `interface MyCustomCache<K extends string> extends CacheLayer<K> {}` is the type of a cache where the key must be a string.

Here's a very basic cache that only caches the first 100 items:

```ts
class MyCache {
  #cache = new Map<string, unknown>();

  get(id: string) {
    if(this.#cache.has(id)) {
      return this.#cache.get(id);
    }
  }

  cache(id: string, value: unknown) {
    if(this.#cache.size > 100) {
      // Don't cache
    } else {
      this.#cache.set(id, value);
    }
  }
}
```

**Comment by okikio** [comment]
Looks pretty good

**Comment by matthewp** [comment]
Created a video that goes over the proposal and explains a bit: https://vimeo.com/795022419/fe90a97420

**Reply by louiss0** [reply]
I have noticed that this is the only implementation of loaders that does not involve the `export` keyword. Since you guys plan on releasing a fetch loader could there be an axios one? And since you guys have decided to use zod as a part of the framework. How about a Zod integration?

**Reply by matthewp** [reply]
The API is low-level on purpose, so that if you wanted to create an Axios version you could do so. If you look at the implementation of `fetch` it's very small: https://github.com/withastro/astro/blob/b68981f23a6f7ea0437938682ffe07b189bc4d49/packages/loaders/src/fetch.ts

Using Zod for schema validation is in the Goals section. That's not present in the examples, I'll add it.

**Reply by matthewp** [reply]
I've added an example which shows using a schema.

**Comment by louiss0** [comment]
This version of loaders seems strange to me. I know that in most frameworks each involve the export keyword. They must be executed in a certain way and in Qwik they are implemented as middleware. How does this version work? **Do you have to call them in an astro server script?**. Do they even have to be executed only on the server? That's how loaders work typically. What are the rules regarding `loaders` in Astro.

**Reply by matthewp** [reply]
If you're thinking of Remix, those loaders are solving a different problem from these. This isn't try to solve server/client data hand-off. This is more similar to react-query, but lower level. It's really just an object that caches data on the server, so that you don't get data differences during the or hit rate limits in dev when working with an external API.

**Reply by louiss0** [reply]
Ok so they are just functions made to cache requests. I thought there would be a little magic. Is there going to be docs on how to create implementation for different libs like `axios`. A  `createFetch()` but for axios.

**Reply by matthewp** [reply]
Yes, that's all they are. Does this make them seem less useful to you?

Yes, we would definitely have extensive documentation if this proposal is accepted.

**Reply by louiss0** [reply]
> Yes, that's all they are. Does this make them seem less useful to you?
> 
> Yes, we would definitely have extensive documentation if this proposal is accepted.

Ok then good luck with your proposal. I was just curious about them.
